contracts/implementation/SMAOracle.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.7;
+import "../interfaces/IOracleWrapper.sol";
+import "../interfaces/IPriceObserver.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "prb-math/contracts/PRBMathSD59x18.sol";
+
+import "hardhat/console.sol";
+
+contract SMAOracle is Ownable, IOracleWrapper {
+    using PRBMathSD59x18 for int256;
+
+    /// Price oracle supplying the spot price of the quote asset
+    address public override oracle;
+
+    /// Price observer providing the SMA oracle with historical pricing data
+    address public observer;
+
+    /// Current SMA price
+    int256 public price;
+
+    /// Number of periods to use in calculating the SMA (`k` in the SMA equation)
+    uint256 public periods;
+
+    int256 public scaler;
+    uint256 public constant MAX_DECIMALS = 18;
+
+    constructor(
+        address _spotOracle,
+        uint256 _spotDecimals,
+        address _observer,
+        uint256 _periods
+    ) {
+        require(_spotOracle != address(0) && _observer != address(0), "SMA: Null address forbidden");
+        require(_periods > 0 && _periods <= IPriceObserver(_observer).capacity(), "SMA: Out of bounds");
+        require(_spotDecimals <= MAX_DECIMALS, "SMA: Decimal precision too high");
+        periods = _periods;
+        setOracle(_spotOracle);
+        setObserver(_observer);
```

Comment:
Should we somehow enforce that the `PriceObserver` for this SMA oracle is this instance of `SMAOracle`? Because functions like `add` and `clear` obviously require the caller to be owner, and given the `SMAOracle` is calling `add` (but not clear though), it would make sense to me that this would only be deployed successfully if the owner was this `SMAOracle` to ensure it will be able to function.